src/rust/src/wallet.rs

```diff
@@ -130,38 +159,182 @@ pub enum WalletError {
+    pub fn is_checkpointed(&self) -> bool {
+        self.last_checkpoint.is_some()
+    }
+
+    /// Rewinds the note commitment tree to the given height, removes notes and
+    /// spentness information for transactions mined in the removed blocks, and returns
+    /// the number of blocks by which the tree has been rewound if successful. Returns  
+    /// `RewindError` if not enough checkpoints exist to execute the full rewind
+    /// requested. If the requested height is greater than or equal to the height of the
+    /// latest checkpoint, this returns `Ok(0)` and leaves the wallet unmodified.
+    pub fn rewind(&mut self, to_height: BlockHeight) -> Result<u32, RewindError> {
+        if let Some(checkpoint_height) = self.last_checkpoint {
+            if to_height >= checkpoint_height {
+                return Ok(0);
+            }
+
+            let blocks_to_rewind = <u32>::from(checkpoint_height) - <u32>::from(to_height);
+
+            // if the rewind length exceeds the number of checkpoints we have stored,
+            // return failure.
+            let checkpoint_count = self.witness_tree.checkpoints().len();
+            if blocks_to_rewind as usize > checkpoint_count {
+                return Err(RewindError::InsufficientCheckpoints(checkpoint_count));
+            }
+
+            for _ in 0..blocks_to_rewind {
+                // we've already checked that we have enough checkpoints to fully
+                // rewind by the requested amount.
+                assert!(self.witness_tree.rewind());
+            }
+
+            // reset our last observed height to ensure that notes added in the future are
+            // from a new block
+            let last_observed = LastObserved {
+                block_height: checkpoint_height - blocks_to_rewind,
+                block_tx_idx: None,
+            };
+
+            // retain notes that correspond to transactions that are not "un-mined" after
+            // the rewind
+            let to_retain: BTreeSet<_> = self
+                .wallet_received_notes
+                .iter()
+                .filter_map(|(txid, n)| {
+                    if n.tx_height
+                        .map_or(true, |h| h <= last_observed.block_height)
+                    {
+                        Some(*txid)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            self.spent_notes.retain(|_, v| to_retain.contains(&v.txid));
+            self.nullifiers.retain(|_, v| to_retain.contains(&v.txid));
+            for (txid, n) in self.wallet_received_notes.iter_mut() {
+                // Erase block height information for any received notes
+                // that have been un-mined by the rewind.
+                if !to_retain.contains(txid) {
+                    n.tx_height = None;
+                }
+            }
+            self.last_observed = Some(last_observed);
```

Comment:
You're correct, this was missing. Originally I had been relying upon the update to `last_observed`, but that doesn't actually work properly because there will be gaps between the blocks that contain Orchard bundles. Now fixed.